src/lib/krb5/krb/pac.c

```diff
@@ -556,64 +557,75 @@ verify_checksum(krb5_context context, const krb5_pac pac, uint32_t buffer_type,
     return ret ? ret : (valid ? 0 : KRB5KRB_AP_ERR_MODIFIED);
 }
 
+/* Per MS-PAC 2.8.3, tickets encrypted to TGS and password change principals
+ * should not have ticket signatures. */
+krb5_boolean
+k5_pac_should_have_ticket_signature(krb5_const_principal sprinc)
+{
+    if (IS_TGS_PRINC(sprinc))
+        return FALSE;
+    if (sprinc->length == 2 && data_eq_string(sprinc->data[0], "kadmin") &&
+        data_eq_string(sprinc->data[1], "changepw"))
+        return FALSE;
+    return TRUE;
+}
+
 static krb5_error_code
-verify_server_checksum(krb5_context context, const krb5_pac pac,
-                       const krb5_keyblock *server)
+verify_pac_checksums(krb5_context context, const krb5_pac pac,
+                     krb5_boolean expect_full_checksum,
+                     const krb5_keyblock *server, const krb5_keyblock *privsvr)
 {
     krb5_error_code ret;
-    krb5_data copy;             /* PAC with zeroed checksums */
+    krb5_data copy, server_checksum;
 
+    /* Make a copy of the PAC with zeroed out server and privsvr checksums. */
     ret = krb5int_copy_data_contents(context, &pac->data, &copy);
     if (ret)
         return ret;
-
-    /* Zero out both checksum buffers */
     ret = zero_signature(context, pac, KRB5_PAC_SERVER_CHECKSUM, &copy);
     if (ret)
         goto cleanup;
     ret = zero_signature(context, pac, KRB5_PAC_PRIVSVR_CHECKSUM, &copy);
     if (ret)
         goto cleanup;
 
-    ret = verify_checksum(context, pac, KRB5_PAC_SERVER_CHECKSUM, server,
-                          KRB5_KEYUSAGE_APP_DATA_CKSUM, &copy);
-
-cleanup:
-    free(copy.data);
-    return ret;
-}
+    if (server != NULL) {
+        /* Verify the server checksum over the PAC copy. */
+        ret = verify_checksum(context, pac, KRB5_PAC_SERVER_CHECKSUM, server,
+                              KRB5_KEYUSAGE_APP_DATA_CKSUM, &copy);
+    }
 
-static krb5_error_code
-verify_kdc_checksum(krb5_context context, const krb5_pac pac,
-                    const krb5_keyblock *privsvr)
-{
-    krb5_error_code ret;
-    krb5_data server_checksum;
+    if (privsvr != NULL && expect_full_checksum) {
+        /* Zero the full checksum buffer in the copy and verify the full
+         * checksum over the copy with all three checksums zeroed. */
+        ret = zero_signature(context, pac, KRB5_PAC_FULL_CHECKSUM, &copy);
+        if (ret)
+            goto cleanup;
+        ret = verify_checksum(context, pac, KRB5_PAC_FULL_CHECKSUM, privsvr,
+                              KRB5_KEYUSAGE_APP_DATA_CKSUM, &copy);
+        if (ret)
+            goto cleanup;
+    } else if (privsvr != NULL) {
```

Comment:
I think this 'else' should be removed and we should check both signatures,
I know this is not strictly needed, but it will catch bugs early and seems to
be more consistent.

Reply:
My reasoning is that the privsvr checksum is vestigial when the full checksum is present and maybe MS will want to phase it out at some point.

We could tolerate a missing privsvr checksum but check it if one is there, if that seems better to you.

Reply:
Yes, that would be better.

Reply:
The code for this got kind of messy (I realised that the zero_checkum() call for the privsvr checksum would also throw an error if it is missing), so I'm just going to put it back the way it was and continue requiring the privsvr checksum to be present.